genomeGraph/genomeGraph.go

```diff
@@ -48,45 +49,61 @@ type Edge struct {
 func Read(filename string) *GenomeGraph {
 	simpleReader := fileio.NewByteReader(filename)
 	genome := EmptyGraph()
-	var currNode, homeNode, destNode *Node
 	var weight float32
 	var line string
 	var words []string = make([]string, 0, 2)
-	var seqIdx, homeNodeIdx, destNodeIdx uint32
+	var nodeId, homeNodeIdx, destNodeIdx uint32
+	var homeNode, destNode *Node
 	var i int
 
 	for reader, done := fileio.ReadLine(simpleReader); !done; reader, done = fileio.ReadLine(simpleReader) {
 		line = reader.String()
 		switch true {
 		case strings.HasPrefix(line, ">"):
-			seqIdx = common.StringToUint32(line[1:])
-			currNode = &Node{Id: seqIdx}
-			AddNode(genome, currNode)
+			nodeId = common.StringToUint32(line[1:])
+			AddNode(genome, &Node{Id: nodeId})
 		case strings.Contains(line, "\t"):
 			words = strings.Split(line, "\t")
 			homeNodeIdx = common.StringToUint32(words[0])
-			homeNode = genome.Nodes[homeNodeIdx]
+			homeNode = &genome.Nodes[homeNodeIdx]
 			if len(words) > 2 {
 				for i = 1; i < len(words); i += 2 {
 					weight = common.StringToFloat32(words[i])
 					destNodeIdx = common.StringToUint32(words[i+1])
-					destNode = genome.Nodes[destNodeIdx]
+					destNode = &genome.Nodes[destNodeIdx]
 					AddEdge(homeNode, destNode, weight)
 				}
 			}
 		default:
-			genome.Nodes[seqIdx].Seq = append(genome.Nodes[seqIdx].Seq, dna.ByteSliceToDnaBases(reader.Bytes())...)
+			genome.Nodes[nodeId].Seq = append(genome.Nodes[nodeId].Seq, dna.ByteSliceToDnaBases(reader.Bytes())...)
 		}
 	}
-	for i = 0; i < len(genome.Nodes); i++ {
-		genome.Nodes[i].SeqTwoBit = dnaTwoBit.NewTwoBit(genome.Nodes[i].Seq)
+	for i = range genome.Nodes {
+		if len(genome.Nodes[i].Seq) != 0 {
+			genome.Nodes[i].SeqTwoBit = dnaTwoBit.NewTwoBit(genome.Nodes[i].Seq)
+		}
 	}
 	return genome
 }
 
-// AddNode will append a new Node to a slice of nodes in GenomeGraph
-func AddNode(g *GenomeGraph, n *Node) {
-	g.Nodes = append(g.Nodes, n)
+// AddNode will add the values in n to the graph at the index of n.Id
+// A pointer to the new location of the node (inside the graph) is returned.
+func AddNode(g *GenomeGraph, n *Node) *Node {
```

Comment:
I sort of understood this return from your description. Happy to approve, but could you talk about this in our next code review?

Reply:
The idea right now is to have Node id given in the input and then never changed.  So if they give us 0, 7, 11 as the node ids, we will respect that.  It is how we can connect the output to the input.  By not re-numbering them, if our software says something maps to node 11, the user will know what that is.  If we re-name 11 to node 2, we would need to store more data so we could change the 2 back to an 11 at the last second.

Reply:
Ahhh thinking about usage, this would be great if we are aligning graphs or making subgraphs right?

Reply:
Yes, that would be the big advantage here.  It would allow graphs to be split apart and merged back together.  It could also allow graphs to be aligned (which could be thought of as putting two sub-graph back together when we are unsure how they were originally split apart).